Mathlib/Order/Interval/Set/OrdConnectedLinear.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2025 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Bhavik Mehta, Oliver Nash
+-/
+import Mathlib.Data.Fintype.Order
+import Mathlib.Data.Nat.Lattice
+import Mathlib.Data.Int.ConditionallyCompleteOrder
+import Mathlib.Data.Int.Interval
+import Mathlib.Data.Int.SuccPred
+
+/-!
+# Order-connected subsets of linear orders
+
+In this file we provide some results about order-connected subsets of linear orders, together with
+some convenience lemmas for characterising closed intervals in certain concrete types such as `ℤ`,
+`ℕ`, and `Fin n`.
+
+## Main results:
+ * `Set.ordConnected_iff_disjoint_Ioo_empty`: a characterisation of `Set.OrdConnected` for
+   locally-finite linear orders.
+ * `Set.Nonempty.ordConnected_iff_of_bdd`: a characterisation of closed intervals for locally-finite
+   conditionally complete linear orders.
+ * `Set.Nonempty.ordConnected_iff_of_bdd'`: a characterisation of closed intervals for
+   locally-finite complete linear orders (convenient for `Fin n`).
+ * `Set.Nonempty.eq_Icc_iff_nat`: characterisation of closed intervals for `ℕ`.
+ * `Set.Nonempty.eq_Icc_iff_int`: characterisation of closed intervals for `ℤ`.
+-/
+
+variable {α : Type*} {I : Set α}
+
+lemma Set.Nonempty.ordConnected_iff_of_bdd
+    [ConditionallyCompleteLinearOrder α] [LocallyFiniteOrder α]
+    (h₀ : I.Nonempty) (h₁ : BddBelow I) (h₂ : BddAbove I) :
+    I.OrdConnected ↔ I = Icc (sInf I) (sSup I) :=
+  have h₄ : I.Finite := h₁.finite_of_bddAbove h₂
+  ⟨fun _ ↦ le_antisymm (subset_Icc_csInf_csSup h₁ h₂)
+    (I.Icc_subset (h₀.csInf_mem h₄) (h₀.csSup_mem h₄)), fun h₃ ↦ h₃ ▸ ordConnected_Icc⟩
+
+/-- A version of `Set.Nonempty.ordConnected_iff_of_bdd` for complete linear orders, such as `Fin n`,
+in which the explicit boundedness hypotheses are not necessary. -/
+lemma Set.Nonempty.ordConnected_iff_of_bdd' [CompleteLinearOrder α] [LocallyFiniteOrder α]
```

Comment:
Now that I think about it, this might work for `OrderBot + OrderTop + ConditionallyCompleteLinearOrder + LocallyFiniteOrder` with exactly the same proof?

Reply:
Done!